crates/ruma-identifiers/src/matrix_uri.rs

```diff
@@ -28,35 +28,112 @@ const TO_ENCODE: &AsciiSet = &CONTROLS
     .add(b';')
     .add(b'=');
 
-/// A reference to a user, room or event.
+/// All Matrix Identifiers that can be represented as a Matrix URI.
+#[derive(Debug, PartialEq, Eq)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+pub enum MatrixId {
+    /// A room ID.
+    Room(Box<RoomId>),
+
+    /// A room alias.
+    RoomAlias(Box<RoomAliasId>),
+
+    /// A user ID.
+    User(Box<UserId>),
+
+    /// An event ID.
+    Event { room_id: Box<RoomOrAliasId>, event_id: Box<EventId> },
+}
+
+impl MatrixId {
+    /// Construct a string with sigils from `self`.
+    ///
+    /// The identifiers will start with a sigil and be percent encoded.
+    ///
+    /// For events, the room ID or alias and the event ID will be separated by
+    /// a slash.
+    pub(crate) fn to_string_with_sigil(&self) -> String {
+        match self {
+            Self::Room(room_id) => percent_encode(room_id.as_bytes(), TO_ENCODE).to_string(),
+            Self::RoomAlias(room_alias) => {
+                percent_encode(room_alias.as_bytes(), TO_ENCODE).to_string()
+            }
+            Self::User(user_id) => percent_encode(user_id.as_bytes(), TO_ENCODE).to_string(),
+            Self::Event { room_id, event_id } => [
+                percent_encode(room_id.as_bytes(), TO_ENCODE).to_string(),
+                percent_encode(event_id.as_bytes(), TO_ENCODE).to_string(),
+            ]
+            .join("/"),
```

Comment:
This could be a `format!("{}/{}")` call, then you shouldn't need the two `to_string()` calls.